### PR TITLE
MueLu: Aggregate quality estimate

### PR DESCRIPTION
**Pull request by eepperly**
@trilinos/muelu

## Description
A method for estimating qualities of the aggregates generated by MueLu for use as a
diagnostic tool was implemented. While originally proposed by Napov and Notay (2012)
as a method for tailoring aggregate generation in the context of unsmoothed
aggregation, the quality metric has also been found to be instructive in diagnosing
poorly made aggregates in the context of smoothed aggregation as well. The method
has been implemented in a new "AggregateQualityEstimateFactory", and has been
integrated into the XML interface.

## How Has This Been Tested?
Unit tests were written to test the new functionality and interface tests were written 
to test the new functionality with the XML interface. New functionality was tested by
comparing against a MATLAB implementation of the same functionality. All new and
existing MueLu unit tests passed on a Linux machine.

## Checklist

- [ ] My commit messages mention the appropriate GitHub issue numbers.
- [x] My code follows the code style of the affected package(s).
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [code contribution guidelines](../blob/master/CONTRIBUTING.md) for this project.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] No new compiler warnings were introduced.
- [ ] These changes break backwards compatibility.

### CODE REVIEW

**Comment by mhoemmen** [comment]
@trilinos/muelu

File: packages/muelu/src/Graph/UncoupledAggregation/MueLu_UncoupledAggregationFactory_def.hpp [hunk 1]
```diff
@@ -50,6 +50,7 @@
 
 #include <Xpetra_Map.hpp>
 #include <Xpetra_Vector.hpp>
+#include <Xpetra_MultiVector.hpp>
```
Comment (lucbv) [inline packages/muelu/src/Graph/UncoupledAggregation/MueLu_UncoupledAggregationFactory_def.hpp]: I think that "Xpetra_Vector.hpp" inludes "Xpetra_MultiVector.hpp" so there is no need for this header

File: packages/muelu/src/Graph/UncoupledAggregation/MueLu_UncoupledAggregationFactory_def.hpp [hunk 2]
```diff
@@ -70,6 +71,7 @@
 #include "MueLu_Aggregates.hpp"
 #include "MueLu_MasterList.hpp"
 #include "MueLu_Monitor.hpp"
+#include "MueLu_FactoryManager.hpp"
```
Comment (lucbv) [inline packages/muelu/src/Graph/UncoupledAggregation/MueLu_UncoupledAggregationFactory_def.hpp]: This file is also already included by "MueLu_UncoupledAggreationFactory_decl.hpp".

File: packages/muelu/src/Misc/MueLu_AggregateQualityEstimateFactory_decl.hpp [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+      void Build(Level & currentLevel) const;
+      void BuildP(Level & currentLevel) const;
+
+      //@}
+
+      //! @name Utility method to convert aggregate data to a convenient format.
+      //@{
+
+      //! Build aggregate quality esimates with this factory.
+      static void ConvertAggregatesData(RCP<const Aggregates> aggs, Teuchos::ArrayRCP<LO>& aggSortedVertices, Teuchos::ArrayRCP<LO>& aggsToIndices, Teuchos::ArrayRCP<LO>& aggSizes);
```
Comment (lucbv) [inline packages/muelu/src/Misc/MueLu_AggregateQualityEstimateFactory_decl.hpp]: More of a comment than a review: `Teuchos::Array`, `Teuchos::ArrayView`, `Teuchos::ArrayRCP` and `Teuchos::RCP` and a few other related and useful classes are imported in the 'muelu' namespace, so no need to add the `Teuchos` prefix. See this file: `muelu/src/Headers/MueLu_ConfigDefs.hpp`

**Comment by jhux2** [comment]
@lucbv wrote:
> Also it do we need to generate the Stokhos class associated with this one?
I'm not sure.  Can we let the PR robot tell us?

**Comment by eepperly** [comment]
@jhux2 @lucbv Committed changes to address the suggested edits

File: packages/muelu/doc/UsersGuide/masterList.xml [hunk 4]
```diff
@@ -607,6 +616,42 @@
...
+      <default>100.0</default>
+      <description>Maximum quality where aggregate is considered acceptable</description>
+      <comment-ML>parameter not existing in ML</comment-ML>
+    </parameter>
+
+    <parameter>
+      <name>aggregate qualities: file output</name>
+      <type>bool</type>
+      <default>false</default>
+      <description>Maximum quality where aggregate is considered acceptable</description>
```
Comment (mayrmt) [inline packages/muelu/doc/UsersGuide/masterList.xml]: Description seems to be wrong.

Comment (eepperly) [reply packages/muelu/doc/UsersGuide/masterList.xml]: Thanks for the catch. Made exactly the mistake you suggested. Should be fixed now

**Comment by eepperly** [comment]
> Also it do we need to generate the Stokhos class associated with this one?

@lucbv I ran the `gen_cppFiles.sh` in the `stokhos/src/muelu/explicit_instantiation/` folder and the following five factories (and `MueLu_AggregateQualityEstimateFactory.cpp`) were newly generated by the script:

```#       MueLu_AggregationStructuredAlgorithm.cpp
#       MueLu_BlackBoxPFactory.cpp
#       MueLu_HybridAggregationFactory.cpp
#       MueLu_InterfaceAggregationAlgorithm.cpp
#       MueLu_UncoupledIndexManager.cpp
```

Should I commit them?

**Comment by prwolfe** [comment]
This has been inactive for over 11 months.  Should we close this PR or is it still being pursued? The last review seemed ok with it and then it's just stalled.

**Comment by jhux2** [comment]
Sorry for the delay.  @csiefer2 and @lucbv recently expressed interest, let's see what they say.

**Comment by csiefer2** [comment]
We committed a modified version of this via another PR.  Thanks for the hard work @eepperly !

**Pull request closed without merging**